lib/compat.php

```diff
@@ -116,6 +116,66 @@ function gutenberg_wpautop( $content ) {
 remove_filter( 'the_content', 'wpautop' );
 add_filter( 'the_content', 'gutenberg_wpautop', 6 );
 
+/**
+ * Add `srcset` to the image meta `sizes` array in the REST API response.
+ *
+ * @param WP_REST_Response $response Response object.
+ * @param WP_Post $attachment Attachment post object.
+ * @return WP_REST_Response Response object.
+ */
+function gutenberg_rest_prepare_attachment( $response, $attachment ) {
+	if ( ! isset( $response->data['media_type'] ) ) {
+		return $response;
+	}
+
+	if ( $response->data['media_type'] === 'image' && ! empty( $response->data['media_details']['sizes'] ) ) {
+		$meta = wp_get_attachment_metadata( $attachment->ID );
+		$sizes = $response->data['media_details']['sizes'];
+
+		foreach ( $sizes as $size_name => $size ) {
+			$srcset = wp_calculate_image_srcset( array( $size['width'], $size['height'] ), $size['source_url'], $meta, $attachment->ID );
```

Comment:
This is an interesting solution, but I wonder if it would be better to create the equivalent of  `wp_calculate_image_srcset()` as a JS function that could be used on the front end when necessary. I'm not sure what the performance impact is of calculating the `srcset` of every possible image size included in an image's `sizes` attribute, but I imagine it isn't insignificant.

Reply:
Additionally, for the REST API, we should open a ticket to add this properly via `register_rest_field`.

Reply:
Yeah, looked into making a JS version of `wp_calculate_image_srcset()` however it will lack all the filtering we currently do in PHP. Also it gets pretty complex when dealing with custom image sizes added by themes and plugins (we can't even "see" any custom sizes in JS, they are in a PHP global).

There is a performance hit for recalculating the `srcset` for each image size but it is pretty small as we already have the attachment meta cached at that time (that's by far the slowest part when calculating `srcset`). Also, this is done only when editing a post and opening the media modal, i.e. there's no slow down from the user's perspective.

A good optimization would be to calculate only one `srcset` per image, for the "large" size, (or the "full" size if the image is small and "large" doesn't exist). However that leaves out the cases where custom sizes were added by a theme. 

Also, in order to handle this better in JS we may need to pass the `srcset` as an array/object (not as string) where the keys are the "size names" so we can look into it easier.

> Additionally, for the REST API, we should open a ticket to add this

Right. Both of these functions will be "merged" to the appropriate places once this is added to core.

Reply:
Makes sense, thanks for the explanation. I don't think it would be out of the question to add a JS hook to a javascript function, but if the performance hit is negligible, I'm not opposed to including that data in REST responses. 

Perhaps doing one `srcset` per image is something we could improve in a future release in combination with something like "image sets" where we group sets of images from the same crop into an array, where we could have a single `srcset` per array.